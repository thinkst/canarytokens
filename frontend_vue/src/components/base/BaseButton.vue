<template>
  <component
    :is="href ? 'a' : 'button'"
    v-bind="$attrs"
    :class="[
      buttonClass,
      { 'flex flex-row flex-nowrap gap-8 items-center': icon },
      { 'flex-row-reverse': iconPosition === 'right' },
      { 'pointer-events-none': loading },
    ]"
    :href="href"
    :type="!href ? type : null"
  >
    <BaseSpinner
      v-if="loading"
      height="1.5rem"
      :variant="
        variant === ButtonVariantEnum.PRIMARY
          ? ButtonVariantEnum.SECONDARY
          : ButtonVariantEnum.PRIMARY
      "
      class="absolute left-0 right-0 ml-auto mr-auto"
    ></BaseSpinner>
    <font-awesome-icon
      v-if="icon"
      :icon="icon"
      aria-hidden="true"
      :class="loading && 'opacity-30'"
    />
    <span :class="loading && 'opacity-30'">
      <slot></slot>
    </span>
  </component>
</template>

<script setup lang="ts">
import type { PropType } from 'vue';
import { computed } from 'vue';
import type { ButtonVariantType } from './types';

enum ButtonVariantEnum {
  PRIMARY = 'primary',
  SECONDARY = 'secondary',
  TEXT = 'text',
  DANGER = 'danger',
  WARNING = 'warning',
  INFO = 'info',
  LIGHT = 'text-light',
}

type buttonType = 'button' | 'submit' | 'reset';

const props = defineProps({
  variant: {
    type: String as PropType<ButtonVariantType>,
    default: 'primary',
  },
  type: {
    type: String as PropType<buttonType>,
    default: 'button',
  },
  icon: {
    type: String,
    default: null,
  },
  iconPosition: {
    type: String as PropType<'left' | 'right'>,
    default: 'left',
  },
  href: {
    type: String || null,
    default: null,
  },
  loading: {
    type: Boolean,
    default: false,
  },
});

const buttonClass = computed(() => {
  switch (props.variant) {
    case ButtonVariantEnum.PRIMARY:
      return 'primary base-button';
    case ButtonVariantEnum.SECONDARY:
      return 'secondary base-button';
    case ButtonVariantEnum.TEXT:
      return 'text base-button';
    case ButtonVariantEnum.DANGER:
      return 'text danger base-button';
    case ButtonVariantEnum.WARNING:
      return 'text warning base-button';
    case ButtonVariantEnum.INFO:
      return 'text info base-button';
    case ButtonVariantEnum.LIGHT:
      return 'text-light base-button';
    default:
      return 'primary base-button';
  }
});
</script>

<style scoped>
.base-button {
  @apply w-fit relative font-semibold rounded-full px-16 py-8 disabled:pointer-events-none transition duration-100;
}

.primary {
  @apply bg-green hover:bg-green-300 disabled:bg-grey-200 disabled:shadow-solid-shadow-grey disabled:border-grey-300 disabled:text-grey-400 active:shadow-none active:top-[0.15rem] focus-visible:outline-0 focus-visible:border-green-800 focus:bg-green-300 focus:border-green-800 focus:outline-0 text-white border shadow-solid-shadow-green border-green-800;
}

.secondary {
  @apply bg-white hover:bg-green-50 hover:text-green-500 focus:text-green-500 disabled:bg-grey-200 disabled:shadow-solid-shadow-grey disabled:border-grey-300 disabled:text-grey-400 active:shadow-none active:top-[0.15rem] focus-visible:outline-0 focus:bg-green-100 focus:border-green-200 focus:outline-0 text-green-600 border shadow-solid-shadow-green-300 border-green-200;
}

.text {
  @apply hover:text-green-500 focus:text-green-800 text-grey-500;
}

.text-light {
  @apply hover:text-green-800 focus:text-green-800 text-white;
}

.danger {
  @apply hover:text-red focus:text-red text-red-500;
}

.warning {
  @apply hover:text-yellow focus:text-yellow text-yellow-700;
}

.info {
  @apply hover:text-blue focus:text-blue text-blue-700;
}
</style>
