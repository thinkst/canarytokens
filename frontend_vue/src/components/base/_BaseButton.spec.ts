import { mount, VueWrapper } from '@vue/test-utils'
import BaseButton from './_BaseButton.vue'

let wrapper: VueWrapper;

describe('BaseButton', () => {
    beforeEach(() => {
        wrapper = mount(BaseButton)
    })

    test('is called', () => {
        expect(wrapper.exists()).toBeTruthy()
    })

    test('render correctly', () => {
        expect(wrapper.html()).toMatchSnapshot()
    })

    test('emits an event when clicked', async () => {
        const onClick = vi.fn()
        wrapper = mount(BaseButton, {
            attrs: { onClick: onClick }
        })
        await wrapper.find('button').trigger('click')
        expect(onClick).toHaveBeenCalledTimes(1)
    })

    test('renders button with primary variant by default', async () => {
        expect(wrapper.classes()).toContain('bg-green')
    })

    test('renders button with secondary variant', async () => {
        wrapper = mount(BaseButton, {
            props: {
            variant: 'secondary'
            },
        })
        expect(wrapper.classes()).toContain('bg-transparent')
    })

    test('renders button with primary variant when prop has a wrong name', async () => {
        wrapper = mount(BaseButton, {
            props: {
            variant: 'wrong-prop-name'
            },
        })
        expect(wrapper.classes()).toContain('bg-green')
    })

    test('renders button with text variant', () => {
        const wrapper = mount(BaseButton, {
        props: { variant: 'text' }
        })
        const expectedClasses = ['font-semibold', 'hover:text-green-800', 'focus:text-green-800', 'text-grey-500', 'rounded-full', 'px-16', 'py-8'];
        expectedClasses.forEach(expectedClass => {
            expect(wrapper.classes()).toContain(expectedClass);
        });
    })

})
