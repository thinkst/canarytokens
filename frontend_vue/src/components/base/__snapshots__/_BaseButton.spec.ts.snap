// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BaseButton > render correctly 1`] = `"<button class="relative bg-green hover:bg-green-300 disabled:pointer-events-none disabled:bg-grey-200 disabled:shadow-solid-shadow-grey disabled:border-grey-300 disabled:text-grey-400 active:shadow-none active:top-[0.15rem] focus-visible:outline-0 focus-visible:border-green-800  focus:bg-green-300 focus:border-green-800 focus:outline-0 text-white border shadow-solid-shadow-green border-green-800 font-semibold rounded-full lg:px-32 lg:py-16 px-16 py-8" type="button"></button>"`;
